### PR TITLE
Follow-up(Edit message)

### LINKED ISSUE #11403: Follow-up(Edit message)
Follow-up to #11374

## **To do 🚧** 
- [x] Mention parsing for the input
- [x] Add hint 
image
- [x] Improve edit history design in MessageButtonsBar
image
- [x] Error handling (showError + keep the edit on)
- [x] Handle capabilities (for Talk Desktop) https://github.com/nextcloud/spreed/pull/11408
  - [x] `delete-messages-unlimited`
  - [x] `edit-messages`
- [ ] Move edit option above "Delete message"
- [ ] https://github.com/nextcloud/spreed/issues/11402

### PR DESCRIPTION
**Pull request by DorraJaouad**
### ☑️ Resolves

* Partially fix #11403

**Remark** 
Mention hint  when editing ( NcNoteCard ) will be shown if the message has `@` initially (before editing it). Otherwise, we will need to compute diff every time the text is edited (not worth it and can be large for long messages) 

### 🖼️ Screenshots / Screencasts

| 🏚️ Before | 🏡 After |
|------------|----------|
| image    | image      |

Newly added 
image

### 🚧 Tasks

- [ ] Visual review
- [ ] Code review

### 🏁 Checklist

- [ ] 🌏 Tested with Chrome, Firefox and Safari or should not be risky to browser differences
- [ ] 🖌️ Design was reviewed, approved or inspired by the design team
- [ ] ⛑️ Tests are included or not possible
- [ ] 📗 User documentation in https://github.com/nextcloud/documentation/tree/master/user_manual/talk has been updated or is not required

### CODE REVIEW

File: src/components/MessagesList/MessagesGroup/Message/MessageButtonsBar/MessageButtonsBar.vue [hunk 1]
```diff
@@ -67,14 +67,11 @@
 					</NcActionText>
 					<!-- Edited message timestamp -->
 					<NcActionButtonGroup v-if="lastEditTimestamp">
```
Comment (nickvergessen) [inline src/components/MessagesList/MessagesGroup/Message/MessageButtonsBar/MessageButtonsBar.vue:69]: Remove? 🙈

File: src/components/MessagesList/MessagesGroup/Message/MessageButtonsBar/MessageButtonsBar.vue [hunk 2]
```diff
@@ -67,14 +67,11 @@
...
+						<NcActionText class="editor-timestamp">
 							<template #icon>
 								<ClockEditOutline :size="16" />
 							</template>
-							{{ lastEditActorDisplayName }}
-						</NcActionText>
-						<NcActionText>
-							{{ editedDateTime }}
+							{{ lastEditActorDisplayName + ' • ' + editedDateTime }}
 						</NcActionText>
```
Comment (nickvergessen) [inline src/components/MessagesList/MessagesGroup/Message/MessageButtonsBar/MessageButtonsBar.vue:75]: My suggestion would be:
```
					<NcActionText v-if="lastEditTimestamp"
						:name="lastEditActorDisplayName">
						<template #icon>
							<ClockEditOutline :size="16" />
						</template>
						{{ editedDateTime }}
					</NcActionText>
```

Bildschirmfoto vom 2024-01-25 17-29-37

File: src/components/MessagesList/MessagesGroup/Message/MessageButtonsBar/MessageButtonsBar.vue [hunk 3]
```diff
@@ -314,12 +309,13 @@ import { useReactionsStore } from '../../../../../stores/reactions.js'
 
 const EmojiIndex = new EmojiIndexFactory(data)
 const supportReminders = getCapabilities()?.spreed?.features?.includes('remind-me-later')
+const canEditMessage = getCapabilities()?.spreed?.features?.includes('edit-messages')
```
[the remark this answers is not in the document]
Comment (Antreesy) [reply src/components/MessagesList/MessagesGroup/Message/MessageButtonsBar/MessageButtonsBar.vue:312]: Same capability could be applied to `NewMessage`: this.messageToEdit could have an `if` check

`MessagesList`, `MessagesGroup` and other visual components should be fine, as old server can't have `lastEdit...` fields to send to the client. But all possible actions in direction `client -> server` should be locked

**Review by Antreesy — COMMENTED** [review]
Nice in general, some technical nitpicks:

File: src/components/MessagesList/MessagesGroup/Message/MessageButtonsBar/MessageButtonsBar.vue (hunk 3 above)
[the remark this answers is not in the document]
Comment (Antreesy) [reply src/components/MessagesList/MessagesGroup/Message/MessageButtonsBar/MessageButtonsBar.vue:314]: Then let's not forget about it)

File: src/stores/__tests__/chatExtras.spec.js [hunk 4]
```diff
@@ -161,4 +162,24 @@ describe('chatExtrasStore', () => {
 			expect(chatExtrasStore.chatInput['token-1']).not.toBeDefined()
 		})
 	})
+
+	describe('text parsing', () => {
```
Comment (Antreesy) [inline src/stores/__tests__/chatExtras.spec.js:166]: You should test functions directly with textParse.spec.js. We could add more examples and tests there later.

In this file, you can call setChatInput() with `{user}` and test that getChatInput() returns `@user`

File: src/components/NewMessage/NewMessage.vue [hunk 5]
```diff
@@ -452,6 +460,10 @@ export default {
 		chatEditInput() {
 			return this.chatExtrasStore.getChatEditInput(this.token)
 		},
+
+		showMentionEditHint() {
+			return this.chatEditInput?.includes('@')
```
[the remark this answers is not in the document]
Comment (DorraJaouad) [reply src/components/NewMessage/NewMessage.vue:465]: hmm probably I should add a leading space or charachter is the first in the text ?

[the remark this answers is not in the document]
Comment (Antreesy) [reply src/components/NewMessage/NewMessage.vue:465]: try this (leading character of followed by space):
```js
  const pattern = /(^|\s)@/
  return pattern.test(this.chatEditInput)
  ```

File: src/utils/textParse.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+ * @param {object} parameters The parameters that contain the mentions
+ * @return {text}
+ */
+function parseMentions(text, parameters) {
+	if (Object.keys(parameters).some(key => key.startsWith('mention'))) {
+		for (const [key, value] of Object.entries(parameters)) {
+			if (value?.type === 'call') {
+				text = text.replace(new RegExp(`{${key}}`, 'g'), '@all')
+			} else if (value?.type === 'user') {
+				text = text.replace(new RegExp(`{${key}}`, 'g'), `@${value.id}`)
```
[the remark this answers is not in the document]
Comment (DorraJaouad) [reply src/utils/textParse.js]: Fixed and also added keyboard shortcut to editing message (last one sent by current user).

**Review by Antreesy — APPROVED** [review]

**Pull request merged**